### PR TITLE
Add support for mixtral model

### PR DESCRIPTION
**Pull request by bhavya01**
- Make the trainer model agnostic and configurable using configs.

- Move sharding configs to model.yaml configs.

- Configure the mesh based on configs. By default, uses only FSDP.

### CODE REVIEW

File: torchprime/torch_xla_models/tests/test_llama.py [hunk 1]
```diff
@@ -65,11 +65,11 @@ def test_forward_our_model_against_hf_model(self):
       )
       torch_xla.sync()
       self.assertTrue(
-        torch.allclose(hf_output.logits, llama_output.logits, atol=1e-6),
+        torch.allclose(hf_output.logits, llama_output[0], atol=1e-6),
```
Comment (tengyifei) [inline torchprime/torch_xla_models/tests/test_llama.py:68]: nit: for readability, we could have something like

```
llama_logits, llama_loss =  model_xla(
  input, labels=input, attention_mask=torch.ones_like(input)
)
```

then we can put `llama_logits` here

File: torchprime/torch_xla_models/train.py [hunk 2]
```diff
@@ -66,18 +63,17 @@ def __init__(
...
+    dcn_mesh_shape = (config.mesh.dcn, 1, 1, 1)
+    ici_mesh_shape = (1, config.mesh.fsdp, config.mesh.tensor, config.mesh.expert)
+    mesh = xs.HybridMesh(
+      ici_mesh_shape=ici_mesh_shape,
+      dcn_mesh_shape=dcn_mesh_shape,
+      axis_names=("dcn", "fsdp", "tensor", "expert"),
     )
+    xs.set_global_mesh(mesh)
+    logger.info(f"Logical mesh shape: {mesh.shape()}")
+    # TODO: Test this for multislice
```
Comment (tengyifei) [inline torchprime/torch_xla_models/train.py:75]: File a bug plz :)

File: torchprime/torch_xla_models/train.py [hunk 3]
```diff
@@ -239,14 +238,32 @@ def train_loop(self):
   @torch_xla.compile(full_graph=True)
   def train_step(self, batch):
     outputs = self.model(**batch)
-    loss = outputs.loss
+    loss = outputs[-1]
```
Comment (tengyifei) [inline torchprime/torch_xla_models/train.py:241]: Maybe `_logits, loss = self.model(**batch)`

File: torchprime/torch_xla_models/train.py [hunk 4]
```diff
@@ -66,18 +63,17 @@ def __init__(
...
-        np.array(range(num_devices)),
-        (num_devices, 1),
-        axis_names=("fsdp", "tensor"),
-      )
-    )
-    logger.info(f"Logical mesh shape: {xs.get_global_mesh().shape()}")
-    self.input_sharding_spec = xs.ShardingSpec(
-      xs.get_global_mesh(), ("fsdp", None), minibatch=True
+    dcn_mesh_shape = (config.mesh.dcn, 1, 1, 1)
+    ici_mesh_shape = (1, config.mesh.fsdp, config.mesh.tensor, config.mesh.expert)
```
Comment (tengyifei) [inline torchprime/torch_xla_models/train.py:67]: Can we check that `num_devices` equals to the product of all these numbers, and raise an error if there's a mismatch? It seems likely someone might forget to override `mesh.fsdp` when running on large clusters.

File: torchprime/torch_xla_models/train.py [hunk 5]
```diff
@@ -175,7 +171,10 @@ def shard_output(output, mesh):
         raise ValueError(
           "Something went wrong, the output of the model shouldn't be `None`"
         )
-      xs.mark_sharding(real_output, mesh, ("fsdp", None, None))
+      # TODO: Find a better way to shard the output. It is expected that the
+      # first dimension of the output is the batch size which is usually sharded
+      # among all the devices except the tensor axis.
+      xs.mark_sharding(real_output, mesh, (("dcn", "fsdp", "expert"), None, None))
```
Comment (tengyifei) [inline torchprime/torch_xla_models/train.py:177]: File a bug plz :)

I guess I didn't know if there could even be a better way. Should this be

```
xs.mark_sharding(real_output, mesh, (("dcn", "fsdp", "expert"), None, "tensor"))
```

? Does that help?

**Review by tengyifei — APPROVED** [review]
This is really nice

File: torchprime/torch_xla_models/train.py (hunk 2 above)
Comment (tengyifei) [reply torchprime/torch_xla_models/train.py:75]: Probably blocked on https://github.com/AI-Hypercomputer/torchprime/issues/52

File: torchprime/torch_xla_models/mixtral/model.py [hunk 6]
```diff
@@ -0,0 +1,1050 @@
...
+    self.layers = nn.ModuleList(
+      [
+        MixtralDecoderLayer(config, layer_idx)
+        for layer_idx in range(config.num_hidden_layers)
+      ]
+    )
+    self.norm = MixtralRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+
+    # Initialize weights and apply final processing
+    # TODO: Remove this
```
[the remark this answers is not in the document]
Comment (bhavya01) [reply torchprime/torch_xla_models/mixtral/model.py:939]: Removed

File: torchprime/torch_xla_models/mixtral/model.py [hunk 7]
```diff
@@ -0,0 +1,1050 @@
...
+    ).repeat_interleave(k)[hidden_states_order]
+    hidden_states_sorted = hidden_states[hidden_states_indices]
+
+    group_sizes = _histogram(top_flat.to(torch.int32), 0, num_experts - 1)
+
+    # Replicated MixtralBlockSparseTop2MLP.forward
+    # Here we just use silu and ignore the configuration given we need to manually write the backward pass.
+    gmm1 = gmm(hidden_states_sorted, w1, group_sizes, tiling=(512, 1024, 1024))
+    gmm3 = gmm(hidden_states_sorted, w3, group_sizes, tiling=(512, 1024, 1024))
+    # Should I save silu activations?
```
[the remark this answers is not in the document]
Comment (bhavya01) [reply torchprime/torch_xla_models/mixtral/model.py:444]: Removed. This is left over from pytorch-tpu. It just says that we hardcode activation function to silu. I don't plan to change this anytime soon.

File: torchprime/torch_xla_models/train.py (hunk 4 above)
Comment (bhavya01) [reply torchprime/torch_xla_models/train.py:67]: Yup, good idea

File: torchprime/torch_xla_models/train.py (hunk 5 above)
Comment (bhavya01) [reply torchprime/torch_xla_models/train.py:177]: I meant that listing all the mesh axes to shard the output of the layer doesn't feel very intuitive. Was wondering if we can make this interface better.

Comment (tengyifei) [reply torchprime/torch_xla_models/train.py]: I see

File: torchprime/torch_xla_models/tests/test_llama.py [hunk 8]
```diff
@@ -86,16 +86,16 @@ def test_forward_torch_xla_against_native(self):
...
     torch_xla.sync()
     self.assertTrue(
-      torch.allclose(
-        llama_output_native.logits, llama_output.logits.to("cpu"), atol=1e-2
-      ),
+      torch.allclose(llama_native_logits, llama_xla_logits.to("cpu"), atol=1e-2),
       "CPU run and XLA run logits are not equal",
     )
     self.assertTrue(
-      torch.allclose(llama_output_native.loss, llama_output.loss.to("cpu"), atol=1e-2),
```
Comment (miladm) [inline torchprime/torch_xla_models/tests/test_llama.py:98]: @bhavya01 @tengyifei is this test part of a training loop?
